pyomo/contrib/doe/DoE_package.py

```diff
@@ -0,0 +1,1812 @@
+        solver = SolverFactory('ipopt')
+        solver.options['linear_solver'] = 'ma57'
+        solver.options['halt_on_ampl_error'] = 'yes'
+        return solver
+
+    def __solve_doe(self, m, fix=False):
+        '''Solve DOE model.
+        If it's a square problem, fix design variable and solve.
+        Else, fix design variable and solve square problem firstly, then unfix them and solve the optimization problem
+
+        m:model
+        fix: if True, solve two times (square first). Else, just solve the square problem
+
+        Return:
+            solver_results: solver results
+        '''
+        ### Solve square problem
+        mod = self.__fix_design(m, self.design_values, fix_opt=fix)
+
+        # if user gives solver, use this solver. if not, use default IPOPT solver
+        solver_result = self.solver.solve(mod,tee=self.tee_opt)
+
+        return solver_result
+
+    def __add_para(self, m, perturb=0):
+        '''
+        For sIPOPT: add parameter perturbation set
+        m: model name
+        self.param_names: perturbation parameter names
+        perturb: which parameter to perturb
+        '''
+        # model parameters perturbation, backward disturb
+        param_do = self.param_v.copy()
+        # perturb parameter
+        param_do[perturb] *= (1-self.step)
+
+        # generate sIPOPT perturbed parameter names
+        param_perturb_names = self.param.copy()
+        for x, xname in enumerate(self.param):
+            param_perturb_names[x] = xname+'_pert'
+        if self.verbose:
+            print('perturb names are:', param_perturb_names)
+
+        self.perturb_names = param_perturb_names
+        if self.verbose:
+            print('Perturbation parameters are set:')
+        for change in range(len(self.perturb_names)):
+            setattr(m, self.perturb_names[change], Param(m.scena, initialize=param_do[change]))
+            if self.verbose:
+                print(self.perturb_names[change], ': ', value(eval('m.'+self.perturb_names[change]+'[0]')))
+        return m
+
+
+class Scenario_generator:
+    def __init__(self, para_dict, formula='central', step=0.001, store=False):
+        '''Generate scenarios.
+        DoE library first calls this function to generate scenarios.
+        For sequential and simultaneous models, call different functions in this class.
+
+        Args:
+        para_dict: a Dict of parameter, keys are names, values are their nominal value. for e.g.,
+                        {'A1': 84.79085853498033, 'A2': 371.71773413976416, 'E1': 7.777032028026428, 'E2': 15.047135137500822}
+        formula: choose from 'central', 'forward', 'backward', None.
+        step: Sensitivity perturbation step size, a fraction between [0,1]. default is 0.001
+        store: if store results
+        '''
+
+        if formula not in ['central', 'forward', 'backward', None]:
+            raise ValueError('Error: undefined formula. Available formulas: central, forward, backward, none.')
```

Comment:
None. Put this in " " and add ' ' around the first three options to be explicit that these are strings. This is ambiguous as written.

Reply:
Do you mean like ["central", "forward", "backward", None]? This None is not a string; it is the python keyword none.

Reply:
Yes.